File: lib/domain/field_cultivation/dtos/field_cultivation_plan_access_snapshot.rb
# frozen_string_literal: true

module Domain
  module FieldCultivation
    module Dtos
      # 圃場栽培と親 CultivationPlan の閲覧・更新可否判定用スナップショット。
      class FieldCultivationPlanAccessSnapshot
        attr_reader :field_cultivation_id, :plan_type_public, :plan_type_private, :plan_user_id

        def initialize(field_cultivation_id:, plan_type_public:, plan_type_private:, plan_user_id:)
          @field_cultivation_id = field_cultivation_id
          @plan_type_public = plan_type_public
          @plan_type_private = plan_type_private
          @plan_user_id = plan_user_id
          freeze
        end

        def plan_type_public?
          plan_type_public
        end

        def plan_type_private?
          plan_type_private
        end
      end
    end
  end
end

File: crates/agrr-domain/src/field_cultivation/dtos/field_cultivation_plan_access_snapshot.rs
//! Ruby: `Domain::FieldCultivation::Dtos::FieldCultivationPlanAccessSnapshot`

/// Ruby: `FieldCultivationPlanAccessSnapshot`
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FieldCultivationPlanAccessSnapshot {
    pub field_cultivation_id: i64,
    pub plan_type_public: bool,
    pub plan_type_private: bool,
    pub plan_user_id: Option<i64>,
}

impl FieldCultivationPlanAccessSnapshot {
    pub fn new(
        field_cultivation_id: i64,
        plan_type_public: bool,
        plan_type_private: bool,
        plan_user_id: Option<i64>,
    ) -> Self {
        Self {
            field_cultivation_id,
            plan_type_public,
            plan_type_private,
            plan_user_id,
        }
    }

    pub fn plan_type_public(&self) -> bool {
        self.plan_type_public
    }

    pub fn plan_type_private(&self) -> bool {
        self.plan_type_private
    }
}
